chore(analytics): Generalize file size analytics

kegluneq opened the pull request:
Move file size analytics code into a reusable module.

See #5312, #5314

Commit:
chore(analytics): Generalize file size analytics

Move file size analytics code into a reusable module.

See #5312, #5314

Commit:
chore(analytics): Build hello_world, check constraints

Create gulp targets to build `hello_world` and check its gzipped size
against size constraints.

See #5312, #5314

Commit:
chore(analytics): Travis job testing Dart payload size

Add a job to the Travis build matrix that checks the size of the
`hello_world` app against targets hard-coded in the gulpfile.

Closes #5314

kegluneq commented:
Yegor volunteered to review this (thanks!), reassigning.

yjbanov commented on modules_dart/payload/hello_world/pubspec.yaml:
```diff
@@ -0,0 +1,25 @@
+    - --show-package-warnings
+    - --trust-type-annotations
+    - --trust-primitives
+    - --enable-experimental-mirrors
```
Do we need experimental mirrors?

kegluneq replied:
Nope, just copied over from `modules/playground`

yjbanov commented on modules_dart/payload/hello_world/web/index.dart:
```diff
@@ -0,0 +1,68 @@
+
+  // You can use the light dom of the <hello-app> tag as temporary content (for
+
+  // example 'Loading...') before the application is ready.
```
I don't think these comments are useful for the purposes of measuring code size

IgorMinar commented on gulpfile.js:
```diff
@@ -1053,35 +1085,9 @@ gulp.task('!bundle.copy', function() {
-  }
+gulp.task('!bundles.js.checksize', function(done) {
+  let reportSize = require('./tools/analytics/reportsize');
+  return reportSize('dist/js/bundle/**', {printToConsole: false, reportAnalytics: true});
```
we want to print to console but only at level2 just like we did before. Can you preserve that please?

kegluneq replied:
Good catch, I completely missed that in the original source. Will reinstate.

IgorMinar commented on gulpfile.js:
```diff
@@ -584,6 +585,37 @@ gulp.task('test.unit.dart', function(done) {
+// Measure in bytes.
+const _DART_PAYLOAD_SIZE_LIMITS = {
+  'uncompressed': 375 * 1024,
+  'gzip level=6': 105 * 1024
```
why level 6? is that what internal servers use? I in js I considered level=2 to be the main level, but I'm happy to adjust if there is evidence that our web servers default to level=6

kegluneq replied:
@yjbanov do you know if this is specified in our internal SLA? I figured "default" is what we would be checking against, and when I compared this value against running `gzip` on my machine `level=6` was the closest value.

yjbanov replied:
Apache uses zlib default, [which is level 6](http://www.zlib.net/manual.html) as of now.

yjbanov replied:
That's also the default gzip, and yes, it level 6 is what I'm using for our SLA.

IgorMinar commented:
the rest looks fine

Commit:
chore(analytics): Include all of angular2 in hello_world

Import dependencies from `angular2/angular2.dart` in `hello_world` to
ensure that any size regressions are caught by our checks.

Update to avoid regressions like #5419.

kegluneq commented:
Updated. Note that this *should* now fail on Travis due to https://github.com/angular/angular/issues/5419 (which Victor is fixing now)

IgorMinar commented:
CI unhappy because you are trying to use `let` in analytics.js which runs non-transpiled. please use `var` instead.

kegluneq commented:
This PR does not update analytics.js, which already used `let`. Is there some reason reportsize.js cannot use `let` & `const` but analytics.js can?

kegluneq commented:
Removing `let` & `const` statements from gulpfile.js seems to have fixed things. Travis is still upset due to dartdoc bug & failing the size check, which should be cleared up soon.

mary-poppins commented:
Merging PR #5394 on behalf of @alxhub to branch [presubmit-alxhub-pr-5394](https://github.com/angular/angular/tree/presubmit-alxhub-pr-5394).
